src/redux/actions/bridge.ts

```diff
@@ -0,0 +1,206 @@
+
+  const updatedAsset = { ...selectedAsset };
+  updatedAsset.balance = {
+    eth: ethBalance,
+    polkadot: polkadotBalance,
+  };
+  dispatch(setSelectedAsset(updatedAsset as TokenData));
+};
+
+// use the token list to instantiate contract instances
+// and store them in redux. We also query the balance of each token for
+// ease of use later
+export const initializeTokens = (tokens: Token[]):
+  ThunkAction<Promise<void>, {}, {}, AnyAction> => async (
+  dispatch: ThunkDispatch<{}, {}, AnyAction>, getState,
+): Promise<void> => {
+  const state = getState() as RootState;
+  const { ethAddress, polkadotApi, polkadotAddress } = state.net;
+  const CoinGeckoClient = new CoinGecko();
+  const priceCurrencies = 'usd';
+
+  if (state.net.web3?.currentProvider) {
+    const web3 = state!.net.web3!;
+    // only include tokens from  current network
+    const tokenList = tokens.filter(
+      (token: Token) => token.chainId
+              === Number.parseInt((web3.currentProvider as any).chainId, 16),
+    );
+
+    // create a web3 contract instance for each ERC20
+    const tokenContractList = tokenList.map(
+      async (token: Token) => {
+        // eth 'tokens' don't have any contract instance
+        let contractInstance = null;
+        let ethBalance = '0';
+        let polkadotBalance = '0';
+        let tokenData;
+        let assetKey = 'ethereum';
+        // price request for eth
+        let priceRequestPromise = CoinGeckoClient.simple.price({
+          ids: assetKey,
+          vs_currencies: priceCurrencies,
+        });
+        let prices = {
+          usd: 0,
+        };
+        // All valid contract addresses have 42 characters ('0x' + address)
+        if (token.address.length === 42) {
+          //   create token contract instance
+          contractInstance = new web3.eth.Contract(
+                ERC20.abi as any,
+                token.address,
+          );
+          // create TokenData for eth API
+          tokenData = {
+            instance: contractInstance,
+            token,
+          } as TokenData;
+
+          // price request for token
+          assetKey = token.address;
+          priceRequestPromise = CoinGeckoClient.simple.fetchTokenPrice({
+            contract_addresses: assetKey,
+            vs_currencies: priceCurrencies,
+          });
+        }
+
+        // fetch token balances
+        try {
+          polkadotBalance = await Polkadot.getEthBalance(
+            polkadotApi!,
+            polkadotAddress!,
+            tokenData,
+          );
+        } catch (e) {
+          console.log('failed reading polkadot balance', e);
+        }
+        try {
+          ethBalance = await EthApi.getTokenBalance(
+            web3,
+            ethAddress!,
+            tokenData,
+          );
+        } catch (e) {
+          console.log('failed reading eth balance', e);
+        }
+
+        // do pricing request
+        const data = await priceRequestPromise;
```

Comment:
can we do this asynchronously/separately to our other queries, rather than blocking on the coingecko api?

Reply:
Yes absolutely, I need to address this so the code is performant with a larger set of tokens